### PR TITLE
Add move constructor and operator to Vector and BlockVector

### PR DESCRIPTION
**Pull request by tamiko**
This pull request adds C++11 move semantic to all Vector and BlockVector
classes (except for the serial variants that will be deprecated).

### CODE REVIEW

File: include/deal.II/lac/petsc_parallel_block_vector.h [hunk 1]
```diff
@@ -100,11 +100,27 @@ namespace PETScWrappers
...
+       * of the vector @p v.
+       *
+       * @note This operator is only available if deal.II is configured with
+       * C++11 support.
+       */
+      BlockVector (BlockVector &&v)
+      {
+        swap(v);
+        // be nice and reset v to zero
+        v.reinit(0, v.get_mpi_communicator(), 0, 0, false);
```
Comment (bangerth) [inline include/deal.II/lac/petsc_parallel_block_vector.h]: This last line may actually be expensive -- it requires communication. It seems like the right thing to do, but at the same time nobody's going to touch the resulting vector. Could we just omit the reinit?

File: include/deal.II/lac/petsc_parallel_vector.h [hunk 2]
```diff
@@ -185,6 +185,26 @@ namespace PETScWrappers
...
+       * Move constructor. Creates a new vector by stealing the internal data
+       * of the vector @p v.
+       *
+       * @note This operator is only available if deal.II is configured with
+       * C++11 support.
+       */
+      Vector (Vector &&v)
+      {
+        swap(v);
+        // be nice and reset v to zero
```
Comment (bangerth) [inline include/deal.II/lac/petsc_parallel_vector.h]: Let's move it out of the class declaration like we do everywhere else.

Comment (bangerth) [inline include/deal.II/lac/petsc_parallel_vector.h]: same everywhere else

File: tests/lac/vector_move_operations.cc [hunk 3]
```diff
@@ -0,0 +1,82 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2012 - 2015 by the deal.II authors
```
Comment (bangerth) [inline tests/lac/vector_move_operations.cc]: you'll need to make this test run only if c++11 is enabled.

Comment (bangerth) [inline tests/lac/vector_move_operations.cc]: ah, forget it -- I forgot that you make this happen in the name of the output file.

File: tests/trilinos/vector_move_operations.with_cxx11=on.output [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (bangerth) [inline tests/trilinos/vector_move_operations.with_cxx11=on.output]: There must be something missing from this output file, right?

**Comment by bangerth** [comment]
Yay -- nice work! Can't wait till we can switch to C++11 altogether!

**Comment by tamiko** [comment]
Everything addressed. There is one more bug in the library we have to take care of:
```
173 template <typename Number>                                                       
174 void BlockVector<Number>::swap (BlockVector<Number> &v)                          
175 {                                                                                
176   Assert (this->n_blocks() == v.n_blocks(),                                      
177           ExcDimensionMismatch(this->n_blocks(), v.n_blocks()));
```
This is not a swap function, this is a mistake :-]

**Comment by bangerth** [comment]
Yay, good work, thanks!

A patch for the BlockVector swap function is pre-approved.

**Comment by Rombur** [comment]
This PR caused more than a hundred tests to fails on both teste and simserver. Can you take a look @tamiko

**Comment by tamiko** [comment]
> This PR caused more than a hundred tests to fails on both teste and
> simserver. Can you take a look @tamiko

Have a look at the currently open pull-request...

**Comment by bangerth** [comment]
In reference to #898.

**Pull request merged**